encoding/string_encode.go

```diff
@@ -1,30 +1,113 @@
 package encoding
 
 import (
-	"errors"
 	"io"
+	"math"
+
+	"github.com/valyala/gozstd"
+	"github.com/xichen2020/eventdb/generated/proto/encodingpb"
+)
+
+var (
+	maxUniqueString = 1 << 16
 )
 
 // StringEncoder encodes string values.
 type StringEncoder interface {
 	// Encode encodes a collection of strings and writes the encoded bytes to the writer.
 	// Callers should explicitly call `Reset` before subsequent call to `Encode`.
-	Encode(writer io.Writer, values ForwardStringIterator) error
+	Encode(writer io.Writer, values RewindableStringIterator) error
 
 	// Reset resets the encoder.
 	Reset()
 }
 
 // StringEnc is a string encoder.
-type StringEnc struct{}
+type StringEnc struct {
+	container []byte
+}
 
 // NewStringEncoder creates a new string encoder.
-func NewStringEncoder(writer io.Writer) *StringEnc { return &StringEnc{} }
+func NewStringEncoder() *StringEnc {
+	return &StringEnc{
+		container: make([]byte, 8),
+	}
+}
 
 // Encode encodes a collection of strings and writes the encoded bytes to the writer.
-func (enc *StringEnc) Encode(writer io.Writer, values ForwardStringIterator) error {
-	return errors.New("not implemented")
+func (enc *StringEnc) Encode(writer io.Writer, values RewindableStringIterator) error {
+	meta := encodingpb.StringMeta{
+		Compression: encodingpb.CompressionType_ZSTD,
+	}
+
+	// Determine whether we want to do table compression
+	mapping := make(map[string]int64)
+	reverseMapping := make([]string, 0)
+	var (
+		idx       int64
+		curr      string
+		maxLength int
+	)
+	for values.Next() {
+		curr = values.Current()
+		if len(mapping) < maxUniqueString+1 {
+			// Only add to mapping if
+			if _, ok := mapping[curr]; !ok {
+				mapping[curr] = idx
+				reverseMapping = append(reverseMapping, curr)
+				idx++
+			}
+		}
+		maxLength = int(math.Max(float64(len(curr)), float64(maxLength)))
+	}
+
+	if values.Err() != nil {
+		return values.Err()
+	}
+
+	// Rewind valuesation.
+	values.Rewind()
+
+	// TODO(bodu): This should take into account the total # of items
+	// at some point. The total count should exceed the # of uniques by a certain
+	// margin to justify table compression.
+	if len(mapping) >= maxUniqueString {
+		meta.Encoding = encodingpb.EncodingType_LENGTH
+		// Min number of bits needed to store length.
+		meta.BytesPerRecord = int64(math.Ceil(MinNumberOfBits(maxLength) / 8))
+	} else {
+		meta.Encoding = encodingpb.EncodingType_DICTIONARY
+		// Min number of bits needed to store dictionary encoded id.
+		meta.BytesPerRecord = int64(math.Ceil(MinNumberOfBits(len(mapping)) / 8))
+	}
+
+	if err := ProtoEncode(&meta, writer); err != nil {
+		return err
+	}
+
+	var compressWriter io.Writer
+	// Compress the bytes.
+	switch meta.Compression {
+	case encodingpb.CompressionType_ZSTD:
+		// TODO(bodu): Figure out a cleaner way to do this.
```

Comment:
You can do
```
var w io.Writer
switch meta.Compression {
  case ...:
    w = gozstd.NewWriter(writer)
    defer w.Close()
  default:
    ...
}
```
?

Also what does closing a `gozstd` writer mean?

Reply:
It finalizes the stream and flushes internal buffers.
https://github.com/valyala/gozstd/blob/master/writer.go#L273

Also, `io.Writer` doesn't implement `io.Closer` so I'm doing that assignment/re-assignment.

Reply:
Sure sounds good.